src/io.jl

```diff
@@ -194,89 +98,102 @@ promote_type2(::Type{Missing}, T::Type{<:Any}) = Union{T, Missing}
-promote_type2(::Type{WeakRefString{UInt8}}, ::Type{Union{WeakRefString{UInt8}, Missing}}) = Union{WeakRefString{UInt8}, Missing}
+promote_type2(::Type{Union{String, Missing}}, ::Type{String}) = Union{String, Missing}
+promote_type2(::Type{String}, ::Type{Union{String, Missing}}) = Union{String, Missing}
 # basic promote type definitions from Base
 promote_type2(::Type{Int64}, ::Type{Float64}) = Float64
 promote_type2(::Type{Float64}, ::Type{Int64}) = Float64
 promote_type2(::Type{Date}, ::Type{DateTime}) = DateTime
 promote_type2(::Type{DateTime}, ::Type{Date}) = DateTime
 # for cases when our current type can't widen, just promote to WeakRefString
-promote_type2(::Type{<:Real}, ::Type{<:Dates.TimeType}) = WeakRefString{UInt8}
-promote_type2(::Type{<:Dates.TimeType}, ::Type{<:Real}) = WeakRefString{UInt8}
-promote_type2(::Type{T}, ::Type{WeakRefString{UInt8}}) where T = WeakRefString{UInt8}
-promote_type2(::Type{Union{T, Missing}}, ::Type{WeakRefString{UInt8}}) where T = Union{WeakRefString{UInt8}, Missing}
-promote_type2(::Type{WeakRefString{UInt8}}, ::Type{T}) where T = WeakRefString{UInt8}
-promote_type2(::Type{WeakRefString{UInt8}}, ::Type{Union{T, Missing}}) where T = Union{WeakRefString{UInt8}, Missing}
+promote_type2(::Type{<:Real}, ::Type{<:Dates.TimeType}) = String
+promote_type2(::Type{<:Dates.TimeType}, ::Type{<:Real}) = String
+promote_type2(::Type{T}, ::Type{String}) where T = String
+promote_type2(::Type{Union{T, Missing}}, ::Type{String}) where T = Union{String, Missing}
+promote_type2(::Type{String}, ::Type{T}) where T = String
+promote_type2(::Type{String}, ::Type{Union{T, Missing}}) where T = Union{String, Missing}
 # avoid ambiguity
-promote_type2(::Type{Any}, ::Type{WeakRefString{UInt8}}) = WeakRefString{UInt8}
-promote_type2(::Type{WeakRefString{UInt8}}, ::Type{Any}) = WeakRefString{UInt8}
-promote_type2(::Type{WeakRefString{UInt8}}, ::Type{WeakRefString{UInt8}}) = WeakRefString{UInt8}
-promote_type2(::Type{WeakRefString{UInt8}}, ::Type{Missing}) = Union{WeakRefString{UInt8}, Missing}
-promote_type2(::Type{Missing}, ::Type{WeakRefString{UInt8}}) = Union{WeakRefString{UInt8}, Missing}
+promote_type2(::Type{Any}, ::Type{String}) = String
+promote_type2(::Type{String}, ::Type{Any}) = String
+promote_type2(::Type{String}, ::Type{String}) = String
+promote_type2(::Type{String}, ::Type{Missing}) = Union{String, Missing}
+promote_type2(::Type{Missing}, ::Type{String}) = Union{String, Missing}
 promote_type2(::Type{Any}, ::Type{Missing}) = Missing
 promote_type2(::Type{Missing}, ::Type{Missing}) = Missing
 
-function detecttype(io, opt::CSV.Options{D}, prevT, levels) where {D}
+function timetype(df::Dates.DateFormat)
+    date = false
+    time = false
+    for token in df.tokens
+        T = typeof(token)
+        if T == Dates.DatePart{'H'}
+            time = true
+        elseif T == Dates.DatePart{'y'} || T == Dates.DatePart{'Y'}
+            date = true
+        end
```

Comment:
If you want to support `ZonedDateTime`s you could add checks for `Dates.DatePart{'z'}` or `Dates.DatePart{'Z'}`